tests/mod_arith/mod-arith-to-arith.mlir

```diff
@@ -223,6 +223,34 @@ func.func @test_lower_mac_vec(%lhs : tensor<4xi8>, %rhs : tensor<4xi8>, %acc : t
 }
 
 
+// -----
+
+// CHECK-LABEL: @test_lower_reduce
+// CHECK-SAME: (%[[ARG:.*]]: [[TENSOR_TYPE:.*]]) -> [[TENSOR_TYPE]] {
+func.func @test_lower_reduce(%arg : tensor<4xi8>) -> tensor<4xi8> {
+  // CHECK: %[[CMOD:.*]] = arith.constant dense<17> : [[TENSOR_TYPE]]
+
+  // CHECK: %[[MOD:.*]] = arith.remsi %[[ARG]], %[[CMOD]] : [[TENSOR_TYPE]]
+  // CHECK: %[[SHIFT:.*]] = arith.addi %[[MOD]], %[[CMOD]] : [[TENSOR_TYPE]]
+  // CHECK: %[[RES:.*]] = arith.remui %[[SHIFT]], %[[CMOD]] : [[TENSOR_TYPE]]
+  %res = mod_arith.reduce %arg { modulus = 17 } : tensor<4xi8>
```

Comment:
I ran into a lot of confusion when I did stuff like this where the modulus was close to some (implicitly signed) container type's upper bound, causing the modulus to be negative. Could we make this attribute typed? Then the user would need to specify `{ modulus = 17 : i32 }`. I think APIntAttr supports this, so it would just be a change to the test code. (If I'm forgetting, maybe `IntegerAttr` or `"::mlir::IntegerAttr"` will do it.

Reply:
I ran into the same confusion now when updating to a bigger modulus (217 which becomes -39 in `i8`). Having added the verify function, the `APInt` is only considered negative when specified as `{ modulus = 217 : i8 }`. For any larger user-specified bit-width it will be considered positive (and default in this case). However, since it is used as an `i8` it will be printed using the `i8` signed interpretation in the IR.

Are you suggesting that we could enforce the unsigned interpretation to be printed in the IR?

Reply:
Not only do I think we can't force the unsigned version to be printed, but when it is interpreted as signed and wraps around to negative, that will actually execute incorrectly (and potentially lower incorrectly), particularly in cases where we use the mod operator (e.g., 217 mod 255 != -39 mod 255, even though 217 = -39 mod 256).

I think we should perhaps have a verifier that ensures the modulus's type is either explicitly unsigned, or else large enough to ensure no 2s complement overflow.